### PR TITLE
Feature/k8s image pipeline tunning

### PR DESCRIPTION
**Pull request by fivejjs**
add docker file with dea_tools and dea_ml

### CODE REVIEW

**Comment by cbur24** [comment]
A few things to address before merging:

- In this document `requirements-odc-static.txt`
these aren't required:
```
otps>=0.3.1   
hdmedians

# Dale's s2cloudmask
# https://github.com/daleroberts/s2cloudmask
s2cloudmask

# need to be installed after to override libs with non-gl ones
opencv-python-headless
opencv-contrib-python-headless--no-binary=\

,hdmedians
```

- and lots of packages in the `requirements.txt` aren't required, though we can probably pair these back progressively until the image is more lean.

- Can you please clean up any files in dea-ml that aren't being used? e.g. feature_layer_default.py , train_feature.py et al.  Starting to get a little messy

**Review by cbur24 — CHANGES_REQUESTED** [review]
See my comments

File: production/dea_ml/docker/requirements-jupyter.txt [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# jupyter things
```
Comment (cbur24) [inline production/dea_ml/docker/requirements-jupyter.txt:1]: If we aren't using this file then let's remove it from the repo

File: production/dea_ml/docker/requirements-odc-static.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# ODC/DEA: these are installed in builder stage
+--extra-index-url="https://packages.dea.ga.gov.au"
+otps>=0.3.1
+hdmedians
+
+eodatasets3
+
```
Comment (cbur24) [inline production/dea_ml/docker/requirements-odc-static.txt:7]: delete

File: production/dea_ml/docker/requirements-odc-static.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# Dale's s2cloudmask
+#  https://github.com/daleroberts/s2cloudmask
+s2cloudmask
+
+# need to be installed after to override libs with non-gl ones
+opencv-python-headless
+opencv-contrib-python-headless
+
+--no-binary=\
+,hdmedians
```
Comment (cbur24) [inline production/dea_ml/docker/requirements-odc-static.txt:21]: delete

File: production/dea_ml/docker/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+structlog
+
+# Scientific Stack
+dask[complete]==2.30.0
+distributed==2.30.0
+dask-image
+dask-ml
+tensorflow==2.3.1
+scipy
+xgboost       # used by Dale's s2cloudmask
```
Comment (cbur24) [inline production/dea_ml/docker/requirements.txt:45]: delete

File: production/dea_ml/docker/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+xarray==0.16.1
+numpy
+zarr
+### plotting
+bokeh==2.1.1
+matplotlib==3.2.1
+descartes
+seaborn
+https://dl.google.com/coral/python/tflite_runtime-2.1.0.post1-cp36-cp36m-linux_x86_64.whl
+
```
Comment (cbur24) [inline production/dea_ml/docker/requirements.txt:65]: delete

File: production/dea_ml/docker/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+mapclassify
+urbanaccess
+pysheds
+https://github.com/ubarsc/rios/archive/rios-1.4.10.tar.gz
+owslib
+rioxarray
+python-geohash
+geohash2
+richdem
+cmocean
```
Comment (cbur24) [inline production/dea_ml/docker/requirements.txt:93]: delete

File: production/dea_ml/docker/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+,regex\
+,ruamel.yaml.clib\
+,yarl\
+,msgpack\
+,multidict\
+,statsmodels\
+,typed_ast\
+,tables\
+,zstandard\
+,richdem\
```
Comment (cbur24) [inline production/dea_ml/docker/requirements.txt:115]: delete

File: production/dea_ml/docker/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+#Install ffmpeg static binary
+RUN apt-get update -y \
+  && apt-get install -y --fix-missing --no-install-recommends xz-utils curl \
+  && rm -rf /var/lib/apt/lists/* \
+  && curl -s -L https://www.johnvansickle.com/ffmpeg/old-releases/ffmpeg-4.2.2-amd64-static.tar.xz > ffmpeg.tar.xz \
+  && echo "c1c7ec5180523db214a7f6f59ea737e8e582570cb884c3ab0c583da4f5165e27" ffmpeg.tar.xz | sha256sum -c - \
+  && tar xvJ --strip-components=1 -C /tmp < ffmpeg.tar.xz \
+  && mkdir -p /usr/local/bin/ \
+  && mv /tmp/ffmpeg /usr/local/bin/ffmpeg \
+  && rm -rf /tmp/* ffmpeg.tar.xz
```
Comment (cbur24) [inline production/dea_ml/docker/Dockerfile:63]: delete

File: production/dea_ml/docker/requirements.txt (hunk 4 above)
Comment (cbur24) [inline production/dea_ml/docker/requirements.txt:45]: delete

File: production/dea_ml/docker/Dockerfile (hunk 8 above)
Comment (fivejjs) [reply production/dea_ml/docker/Dockerfile:63]: is that affect the tif read and write ?

File: production/dea_ml/docker/requirements.txt (hunk 4 above)
Comment (fivejjs) [reply production/dea_ml/docker/requirements.txt:45]: dask-ml is used by predict_xr
scipy could be the dependency of scikit-learn

**Comment by alexgleith** [comment]
You're better off starting with this: https://github.com/opendatacube/datacube-docker/tree/main/statistician and adding things.

File: production/dea_ml/docker/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+USER $nb_user
+RUN echo "Adding odc-dependencies" \
+  && if [ "${PURGE_CACHE_ODC}" != "N" ]; then echo "Forced docker to purge cache for this step!"; fi \
+  && ${py_env_path}/bin/pip install \
+  --no-cache-dir \
+  --requirement=/conf/requirements-odc.txt
+
+# install crop-mask-dependencies dea-tools and dea-ml through github http url
+COPY dea-tools-ml.txt /conf/
+USER $nb_user
```
Comment (alexgleith) [inline production/dea_ml/docker/Dockerfile:139]: This image is not running notebooks.

**Comment by alexgleith** [comment]
Also, it's better to have a PR contain a set of changes that are interdependent, so mixing code changes and a Dockerfile makes this hard.

File: production/dea_ml/docker/Dockerfile (hunk 9 above)
Comment (fivejjs) [reply production/dea_ml/docker/Dockerfile:139]: that is to install only following two packages without running the jupyterhub or jupyterlab
```
# use minty-fresh-sandbox branch
git+https://github.com/digitalearthafrica/deafrica-sandbox-notebooks.git@minty-fresh-sandbox#subdirectory=Tools
# use main branch here
git+https://github.com/digitalearthafrica/crop-mask.git@main#subdirectory=production/dea_ml
```

**Comment by fivejjs** [comment]
> Also, it's better to have a PR contain a set of changes that are interdependent, so mixing code changes and a Dockerfile makes this hard.

File: production/dea_ml/docker/requirements.txt (hunk 4 above)
Comment (cbur24) [reply production/dea_ml/docker/requirements.txt]: Yes, I only intended for xgboost to be deleted, for some reason github shows a larger preview than the lines selected

File: production/dea_ml/docker/Dockerfile (hunk 8 above)
Comment (cbur24) [reply production/dea_ml/docker/Dockerfile:63]: I thought ffmeg is for video and audio files

**Pull request closed without merging**